In [ ]:
import pandas as pd
import numpy as np
import random
import io

# Data import

In [5]:
econdata = pd.read_csv('/content/drive/MyDrive/cursos_platzi/Inferencia/econdata.csv')

In [6]:
econdata.head()

id                  geo_point_2d  \
0   0   19.424781053,-99.1327537959   
1   1  19.4346139576,-99.1413808393   
2   2  19.4340695945,-99.1306348409   
3   3      19.42489472,-99.12073393   
4   4      19.42358238,-99.12451093   

                                           geo_shape   clave_cat  \
0  {"type": "Polygon", "coordinates": [[[-99.1332...  307_130_11   
1  {"type": "MultiPoint", "coordinates": [[-99.14...  002_008_01   
2  {"type": "MultiPoint", "coordinates": [[-99.13...  006_002_12   
3  {"type": "MultiPoint", "coordinates": [[-99.12...  323_102_06   
4  {"type": "MultiPoint", "coordinates": [[-99.12...  323_115_12   

            delegacion perimetro     tipo  \
0           Cuauhtémoc         B  Mercado   
1            Cuautémoc         A    Museo   
2            Cuautémoc         A    Museo   
3  Venustiano Carranza         B    Hotel   
4  Venustiano Carranza         B    Hotel   

                                              nom_id  
0                                        Pino Suárez  
1  Museo Nacional de Arquitectura Palacio de Bell...  
2                                       Santa Teresa  
3                                           Balbuena  
4                                               real

# Sample functions

## 1) Simple random sampling

In [10]:
rand8 = econdata.sample(n=8)
rand8

id                  geo_point_2d  \
224  224  19.4238397797,-99.1371356587   
124  124  19.4297560476,-99.1253498219   
1      1  19.4346139576,-99.1413808393   
172  172  19.4340126845,-99.1357099089   
197  197   19.4399151122,-99.149385962   
168  168   19.4349726565,-99.147766133   
43    43  19.4343713052,-99.1291939496   
52    52  19.4363816901,-99.1293374805   

                                             geo_shape   clave_cat delegacion  \
224  {"type": "MultiPoint", "coordinates": [[-99.13...  001_109_11  Cuautémoc   
124  {"type": "MultiPoint", "coordinates": [[-99.12...  006_052_01  Cuautémoc   
1    {"type": "MultiPoint", "coordinates": [[-99.14...  002_008_01  Cuautémoc   
172  {"type": "MultiPoint", "coordinates": [[-99.13...  001_012_12  Cuautémoc   
197  {"type": "MultiPoint", "coordinates": [[-99.14...  012_153_01  Cuautémoc   
168  {"type": "MultiPoint", "coordinates": [[-99.14...  002_014_23  Cuautémoc   
43   {"type": "MultiPoint", "coordinates": [[-99.12...  006_003_19  Cuautémoc   
52   {"type": "MultiPoint", "coordinates": [[-99.12...  005_124_05  Cuautémoc   

    perimetro   tipo                                             nom_id  
224         B  Hotel                               Costa Azul, SA DE CV  
124         A  Hotel                                           Universo  
1           A  Museo  Museo Nacional de Arquitectura Palacio de Bell...  
172         A  Hotel                                             Cadana  
197         B  Hotel                                               Mina  
168         B  Hotel                                        One Alameda  
43          A  Hotel                                           Valencia  
52          A  Museo                                 Las Constituciones

In [14]:
prop25 = econdata.sample(frac=0.25)
prop25.head()

id                  geo_point_2d  \
100  100  19.4337759404,-99.1378211234   
12    12      19.43990186,-99.14813347   
148  148  19.4388349207,-99.1502578596   
196  196  19.4437017048,-99.1325438818   
11    11  19.4219963186,-99.1437414652   

                                             geo_shape   clave_cat delegacion  \
100  {"type": "MultiPoint", "coordinates": [[-99.13...  001_014_06  Cuautémoc   
12   {"type": "MultiPoint", "coordinates": [[-99.14...  003_079_16  Cuautémoc   
148  {"type": "MultiPoint", "coordinates": [[-99.15...  012_147_13  Cuautémoc   
196  {"type": "MultiPoint", "coordinates": [[-99.13...  004_043_06  Cuautémoc   
11   {"type": "MultiPoint", "coordinates": [[-99.14...  002_103_18  Cuautémoc   

    perimetro   tipo                  nom_id  
100         A  Hotel  Ritz Ciudada de México  
12          B  Hotel                  La Paz  
148         B  Hotel      Miño, S.A. DE C.V.  
196         B  Hotel                  Boston  
11          B  Hotel                    Faro

## 2) Systematic sampling

In [15]:
def sys_sampling(data, step):
  ind = np.arange(0, len(data), step=step)
  sample = data.iloc[ind]
  return sample

In [16]:
systematic_sample = sys_sampling(econdata, 3)
systematic_sample

id                  geo_point_2d  \
0      0   19.424781053,-99.1327537959   
3      3      19.42489472,-99.12073393   
6      6      19.43553422,-99.12324801   
9      9  19.4407152937,-99.1498060057   
12    12      19.43990186,-99.14813347   
..   ...                           ...   
216  216  19.4247697438,-99.1249707246   
219  219  19.4234096295,-99.1438351156   
222  222   19.4397905042,-99.134628735   
225  225      19.43094655,-99.12455418   
228  228  19.4406567734,-99.1394531646   

                                             geo_shape   clave_cat  \
0    {"type": "Polygon", "coordinates": [[[-99.1332...  307_130_11   
3    {"type": "MultiPoint", "coordinates": [[-99.12...  323_102_06   
6    {"type": "MultiPoint", "coordinates": [[-99.12...  318_116_11   
9    {"type": "MultiPoint", "coordinates": [[-99.14...  012_146_22   
12   {"type": "MultiPoint", "coordinates": [[-99.14...  003_079_16   
..                                                 ...         ...   
216  {"type": "Polygon", "coordinates": [[[-99.1250...  323_118_37   
219  {"type": "MultiPoint", "coordinates": [[-99.14...  002_118_03   
222  {"type": "MultiPoint", "coordinates": [[-99.13...  004_063_10   
225  {"type": "MultiPoint", "coordinates": [[-99.12...  323_027_04   
228  {"type": "MultiPoint", "coordinates": [[-99.13...  004_069_01   

              delegacion perimetro     tipo                  nom_id  
0             Cuauhtémoc         B  Mercado             Pino Suárez  
3    Venustiano Carranza         B    Hotel                Balbuena  
6    Venustiano Carranza         B    Hotel    San Antonio Tomatlan  
9              Cuautémoc         B    Hotel                 Detroit  
12             Cuautémoc         B    Hotel                  La Paz  
..                   ...       ...      ...                     ...  
216  Venustiano Carranza         B  Mercado  Lamininas de La Merced  
219            Cuautémoc         B    Hotel          Prado Floresta  
222            Cuautémoc         A    Hotel                  Ladero  
225  Venustiano Carranza         B    Hotel           Gran Veracruz  
228            Cuautémoc         B    Museo        Tequila y Mescal  

[77 rows x 8 columns]

## 3) Stratified sampling

In [19]:
econdata['estratificado'] = econdata['delegacion']+', '+econdata['tipo']
(econdata['estratificado'].value_counts() / len(econdata)).sort_values(ascending=False)

Cuautémoc, Hotel                0.643478
Cuautémoc, Museo                0.156522
Venustiano Carranza, Hotel      0.078261
Cuauhtémoc, Mercado             0.073913
Venustiano Carranza, Mercado    0.047826
Name: estratificado, dtype: float64

Data shows the following proportion

1. Cuautémoc hotels: 64.3%
2. Cuautémoc museums: 15.6%
3. Venustiano Carranza: 7.8%
4. Cuauhtémoc: 7.3%
5. Venustiano Carranza: 4.7%


In [23]:
def strat_sampling(data, strat_col_names, strat_values, strat_prop, random_state=None):
  df_strat = pd.DataFrame(columns=data.columns)
  pos=-1
  for i in range(len(strat_values)):
    pos+=1
    if pos == len(strat_values)-1:
      ration_len = len(data)-len(df_strat)
    else:
      ratio_len = int(len(data)*strat_prop[i])
    
    df_filtered = data[data[strat_col_names]==strat_values[i]]
    df_temp = df_filtered.sample(replace=True, n=ratio_len, random_state=random_state)

    df_strat = pd.concat([df_strat, df_temp])
  return df_strat

In [24]:
strat_values = ['Cuautémoc, Hotel' , 'Cuautémoc, Museo', 'Venustiano Carranza, Hotel', 'Cuauhtémoc, Mercado', 'Venustiano Carranza, Mercado']
strat_prop = [0.643478, 0.156522, 0.078261, 0.073913, 0.047826]
df_strat = strat_sampling(econdata, 'estratificado', strat_values, strat_prop, random_state=42)

In [25]:
df_strat

id                  geo_point_2d  \
164  164  19.4388741511,-99.1413308257   
142  142  19.4263681354,-99.1327278126   
27    27  19.4348360773,-99.1463945583   
168  168   19.4349726565,-99.147766133   
113  113      19.43374405,-99.13550135   
..   ...                           ...   
163  163  19.4265454033,-99.1224859032   
128  128  19.4270781084,-99.1210175514   
77    77  19.4254536359,-99.1226996335   
163  163  19.4265454033,-99.1224859032   
163  163  19.4265454033,-99.1224859032   

                                             geo_shape        clave_cat  \
164  {"type": "MultiPoint", "coordinates": [[-99.14...       003_113_03   
142  {"type": "MultiPoint", "coordinates": [[-99.13...       006_127_14   
27   {"type": "MultiPoint", "coordinates": [[-99.14...       002_016_01   
168  {"type": "MultiPoint", "coordinates": [[-99.14...       002_014_23   
113  {"type": "MultiPoint", "coordinates": [[-99.13...       001_012_13   
..                                                 ...              ...   
163  {"type": "Polygon", "coordinates": [[[-99.1231...       323_063_05   
128  {"type": "Polygon", "coordinates": [[[-99.1214...  323_061_04(123)   
77   {"type": "Polygon", "coordinates": [[[-99.1233...       323_101_01   
163  {"type": "Polygon", "coordinates": [[[-99.1231...       323_063_05   
163  {"type": "Polygon", "coordinates": [[[-99.1231...       323_063_05   

              delegacion perimetro     tipo                   nom_id  \
164            Cuautémoc         B    Hotel             Dos Naciones   
142            Cuautémoc         A    Hotel                    Ambar   
27             Cuautémoc         B    Hotel  Hilton Centro Histórico   
168            Cuautémoc         B    Hotel              One Alameda   
113            Cuautémoc         A    Hotel              San Antonio   
..                   ...       ...      ...                      ...   
163  Venustiano Carranza         B  Mercado                      NaN   
128  Venustiano Carranza         B  Mercado              San Ciprian   
77   Venustiano Carranza         B  Mercado     La Merced Nave Menor   
163  Venustiano Carranza         B  Mercado                      NaN   
163  Venustiano Carranza         B  Mercado                      NaN   

                    estratificado  
164              Cuautémoc, Hotel  
142              Cuautémoc, Hotel  
27               Cuautémoc, Hotel  
168              Cuautémoc, Hotel  
113              Cuautémoc, Hotel  
..                            ...  
163  Venustiano Carranza, Mercado  
128  Venustiano Carranza, Mercado  
77   Venustiano Carranza, Mercado  
163  Venustiano Carranza, Mercado  
163  Venustiano Carranza, Mercado  

[233 rows x 9 columns]